### PR TITLE
When a listening socket is closed, close also all orphans

### PR DESCRIPTION
**Pull request by bernd-edlinger**
Description
-----------

Close all not yet accepted sockets when the listening socket is closed.
Otherwise these sockets become orphans and cannot be accepted any more.
Already accepted sockets must be closed by the application.

Test Steps
-----------
Close a listening socket when there are new connections which are not accepted.
See it is impossible to bind a new server socket on the same port again, since the
orphans occupy the port.

Related Issue
-----------

https://forums.freertos.org/t/issues-with-freertos-closesocket/12195

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: FreeRTOS_Sockets.c
Comment (htibosch) [inline FreeRTOS_Sockets.c:1618]: MISRA doesn't like a `return` in the middle of a function, so could you change it into an else?

**Review by htibosch — APPROVED** [review]
Until now, orphaned sockets would only be removed after a time-out, by the IP-task when `ipconfigTCP_HANG_PROTECTION` is defined.

After this PR, when the listening socket is closed, any pending sockets ( in the SYN phase ) will get closed immediately.

Thanks, Hein

File: FreeRTOS_Sockets.c
Comment (AniruddhaKanhere) [inline FreeRTOS_Sockets.c:1608]: This is a very minor suggestion to keep both the if and else clauses similar to each other.  Why not move this up in the for loop condition above?
`pxIterator = listGET_NEXT( pxIterator );`

i.e., the loop would become:
```
            for( pxIterator = listGET_NEXT( pxEnd );
                 pxIterator != pxEnd; 
                 pxIterator = listGET_NEXT( pxIterator )    )
```

Comment (AniruddhaKanhere) [reply FreeRTOS_Sockets.c:1608]: I don't mind this how it is. I shall approve the PR.

**Review by AniruddhaKanhere — APPROVED** [review]

File: FreeRTOS_Sockets.c
Comment (bernd-edlinger) [reply FreeRTOS_Sockets.c:1608]: Hmm, I think that would not work, because of the 
vSocketClose below I need to do the listGET_NEXT
here, the original form of the loop would access free
memory.

However, I agree that the reason why I wrote it this way
is probably not so clear as I thought it would be, so I should
definitely add a comment here, to prevent such "code-cleanup"
in the future.

Comment (bernd-edlinger) [reply FreeRTOS_Sockets.c]: I've added a comment to hopefully make the code more clear:
"This needs to be done here, before calling vSocketClose."
Thanks for bringing this to may attention.

**Review by AniruddhaKanhere — APPROVED** [review]

File: FreeRTOS_Sockets.c
Comment (AniruddhaKanhere) [reply FreeRTOS_Sockets.c]: Ah! Yes, I did not see the cast. Thanks for explaining that.
I have approved the PR. But it needs to be rebased before it can be merged.

**Comment by AniruddhaKanhere** [comment]
@bernd-edlinger, please allow FreeRTOS members to contribute to your repository. That would allow us to rebase your PR on top of main. This is required since we only allow linear history of commits.
This is outlined in the Contributing document [here](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/blob/main/.github/CONTRIBUTING.md#contributing-via-pull-request). Paraphrasing: "*NOTE: Please make sure the default option (Allow edits from maintainers) is left checked.*"

File: FreeRTOS_Sockets.c
Comment (htibosch) [inline FreeRTOS_Sockets.c:1614]: For me the expression is perfect, but MISRA would like to see:
~~~c
( ( pxOtherSocket->u.xTCP.bits.bPassQueued != 0U ) ||
  ( pxOtherSocket->u.xTCP.bits.bPassAccept != 0U ) )
~~~

**Comment by bernd-edlinger** [comment]
Done.

File: FreeRTOS_Sockets.c
Comment (bernd-edlinger) [reply FreeRTOS_Sockets.c]: okay, done.

**Comment by bernd-edlinger** [comment]
rebased to master and enabled the allow edits checkbox

**Review by AniruddhaKanhere — APPROVED** [review]

**Review by n9wxu — APPROVED** [review]

**Comment by AniruddhaKanhere** [comment]
@bernd-edlinger, thanks a lot for this PR and for your patience! We are working on (reviewing and) merging your other PRs as well. 
Merging the PR now.

**Pull request merged**
